experimental/algorithm/LAGraph_RegularReachability.c

```diff
@@ -0,0 +1,329 @@
+// of paths starting in one of the source nodes. And the second one is the set
+// of the paths accepted by the NFA thus matching the desired restrictions.
+//
+// On algorithm step n the relation between the NFA edges and the graph nodes
+// is bulit. These conditions are held for the pairs in this relation:
+// * The node is reachable by a path of length n from one of the source nodes
+//   in the graph.
+// * The state is reachable by a path of length n from one of the starting
+//   states in the NFA.
+// * These paths have the same labels.
+// The algorithm accumulates these relations. Then it extracts the nodes that
+// are in relation with one of the final states.
+//
+// Full description is available at:
+//   http://arxiv/will/be/here/i/hope
+//
+// Performance considerations: the best performance is shown when the algorithm
+// receives a minimal deterministic finite automaton as an input.
+
+#define LG_FREE_WORK                            \
+{                                               \
+    GrB_free (&frontier) ;                      \
+    GrB_free (&next_frontier) ;                 \
+    GrB_free (&symbol_frontier) ;               \
+    LAGraph_Free ((void **) &A, NULL) ;         \
+    LAGraph_Free ((void **) &AT, NULL) ;        \
+    LAGraph_Free ((void **) &B, NULL) ;         \
+    LAGraph_Free ((void **) &BT, NULL) ;        \
+}
+
+#define LG_FREE_ALL                 \
+{                                   \
+    LG_FREE_WORK ;                  \
+    GrB_free (reachable) ;          \
+}
+
+#include "LG_internal.h"
+#include "LAGraphX.h"
+
+int LAGraph_RegularReachability
+(
+    // output:
+    GrB_Vector *reachable,      // reachable(i) = true if node i is reachable
+                                // from one of the starting nodes
+    // input:
+    LAGraph_Graph *R,           // input non-deterministic finite automaton
+                                // adjacency matrix decomposition
+    size_t nl,                  // total label count, # of matrices graph and
+                                // NFA adjacenc ymatrix decompositions
+    const GrB_Index *QS,        // starting states in NFA
+    size_t nqs,                 // number of starting states in NFA 
+    const GrB_Index *QF,        // final states in NFA
+    size_t nqf,                 // number of final states in NFA 
+    LAGraph_Graph *G,           // input graph adjacency matrix decomposition
+    const GrB_Index *S,         // source vertices to start searching paths
+    size_t ns,                  // number of source vertices
+    char *msg
+)
+{
+
+    //--------------------------------------------------------------------------
+    // check inputs
+    //--------------------------------------------------------------------------
+
+    LG_CLEAR_MSG ;
+
+    GrB_Matrix frontier = NULL ;
+    GrB_Matrix symbol_frontier = NULL ;
+    GrB_Matrix next_frontier = NULL ;
+    GrB_Matrix visited = NULL ;
+    GrB_Vector final_reducer = NULL ;
```

Comment:
Who is it?

Reply:
Resolving, since discussed verbally Semyon Vyacheslavovich is ok.